[Linting] Fix implicit `Optional` type hints

jond01 opened the pull request:
Use Ruff's RUF013 rule:
https://docs.astral.sh/ruff/rules/implicit-optional/

Note: this rule is limited. It does not catch aliases or custom classes, e.g.:

* https://github.com/mlrun/mlrun/blob/8e973600d3bfb8ff6af8cfd2ce21f6a72a0c77a0/mlrun/db/base.py#L88
* https://github.com/mlrun/mlrun/blob/8e973600d3bfb8ff6af8cfd2ce21f6a72a0c77a0/mlrun/db/httpdb.py#L1518

Pyright's [reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType) check catches such errors. We can add it to better cover this check.
See also https://github.com/astral-sh/ruff/issues/14018.

jond01 committed:
Add RUF013 rule to the configuration

liranbg commented:
@jond01 since this was an automated change, please hold until https://github.com/mlrun/mlrun/pull/6615 is merged

assaf758 commented on automation/deployment/deployer.py:
```diff
@@ -48,23 +48,23 @@ class ExcecutionParams:
-        override_mlrun_log_collector_image: str = None,
-        override_mlrun_ui_image: str = None,
-        override_jupyter_image: str = None,
+        registry_secret_name: typing.Optional[str] = None,
```
For brevity, please import `Optional` in modules where it's missing.

jond01 replied:
@assaf758 this change requires a lot of manual work or a clever way to detect imports across many modules.
I think it's ok for now, new code can choose to use `from typing import Optional` instead of `import typing`.

assaf758 replied:
Fair enough.
Please just fix the file conflict.

liranbg commented:
you have the green light to continue

jond01 committed:
[Typing] Fix type hints with `datetime` module instead of the class

Cherry-picked from #6607:
https://github.com/mlrun/mlrun/pull/6607/commits/582256d5d249a402e6e66869cb68538d7b3aec74

jond01 committed:
Fix RUF013 errors

```sh
ruff check --select RUF013 --unsafe-fixes --fix
ruff format
ruff check --fix
```
